### ISSUE TITLE
[artifactory-ha] PostStartHookError - can't create directory '/artifactory_bootstrap/artifactory_extra_conf': Permission denied

### ISSUE THREAD

**Issue by angegar** [jfrog/charts#862]
**Is this a request for help?**: Yes

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG REPORT

**Version of Helm and Kubernetes**: 1.15.10

**Which chart**: 2.4.3 but also present in 2.3.1

**What happened**: I am trying to migrate from 6.18.1

**What you expected to happen**: Migration complete successfully

**How to reproduce it** (as minimally and precisely as possible):
My database is MySQL database. The current chart version installed is 1.5.3.

**Anything else we need to know**:

Here is the symptom :

![image](https://user-images.githubusercontent.com/5628764/80343367-07f8a980-8866-11ea-8497-d949925da831.png)

**Comment by chukka** [comment]
@angegar Thanks for sharing the information. Can you please share the reproducible steps ( values.yaml file helps)

**Comment by angegar** [comment]
Hi @chukka , i am still struggling with this ;)

I see some inconsistency in the primary statefulset. Indeed the command run is `/bin/bash -c ....` whereas the `lifecycle.postStart.exec.command = /bin/sh`. I don't think this is the root cause but probably something to fix.

**Comment by chukka** [comment]
@angegar  sorry for delayed repsonse, That change from `/bin/sh` to `/bin/bash` will be fixed in next chart version, how ever looks like  you are running into different issue, can you share artifactory logs to debug further

**Comment by angegar** [comment]
To be honest i made so many try and fail that it will be hard to redo everything. Now i am looking for a working configuration when this step achieve, i will destroy the environment and rollback to the chart 1.5.3 then reapply the changes.

For now i read this message from the migratio-artifactory-ha : Migration not necessary. artifactory is already 7.3.2 .

I deleted the full stack with `helm delete --purge artifacory`.
I restarted the deployment but it is stuck on primary-0 PodInitializing but I don't find any error log.

**Comment by angegar** [comment]
I forget the chart version is 2.3.1

**Comment by angegar** [comment]
I just deleted the helm release and rerun the deployment pipeline. Now the error symptoms are :

`PostStartHookError: command '/bin/bash -c echo;`

```
Copying bootstrap config from /artifactory_extra_conf to /artifactory_bootstrap
cp: can't create directory '/artifactory_bootstrap/artifactory_extra_conf': Permission denied
```

maybe the uid and gid are different in the latest chart version

**Comment by angegar** [comment]
The persmission denied is coming from : 

![image](https://user-images.githubusercontent.com/5628764/80978641-a072c900-8e26-11ea-9db4-5cb5e323b19c.png)

Which is weird because i always ran artifactory with the defautl uid gid.

I will try to clean the helm release as well as the namespaces and the pvc.

**Comment by angegar** [comment]
Here is a full view on the permissions :

![image](https://user-images.githubusercontent.com/5628764/80982718-fc8c1c00-8e2b-11ea-92a6-a7fafe5ec617.png)

**Comment by chukka** [comment]
@angegar Thanks for the info .This is a known permissions bug (fix  is work in progress - will be available in the next release) .

**Comment by angegar** [comment]
@chukka , Thank you for your answer but i have to move forward. What can i do to make it work.

I restarted from a fresh value.yml file, and it is getting better. However i have now a new issue probably due to the migration;


```
2020-05-05T06:45:46.326Z [jfrt ] [ERROR] [359f1b764cf3d60c] [tifactoryHomeConfigListener:59] [ocalhost-startStop-2] - Failed initializing Home. Version 13 no found.
2020-05-05T06:45:46.327Z [jfrt ] [ERROR] [359f1b764cf3d60c] [actoryContextConfigListener:82] [ocalhost-startStop-2] - Failed initializing Artifactory context: Artifactory home not initialized.
```

Do you know how can i fix this ?

**Comment by chukka** [comment]
@angegar The above  permissions bug impacts boot config files which are placed in following directories  `artifactory_extra_conf` in 6.x and `artifactory_bootstrap` in 7.x , If you don't use these, you are good to upgrade from 6.x to 7.x . Hope it helps

**Comment by angegar** [comment]
Indeed I changed my configuration for : 
```yaml
  customVolumes: |
    - name : logback
      configMap:
        name: logback
  customVolumeMounts: |
      - name: logback
        mountPath: /tmp/artifactory-logback

  # Files to copy to ARTIFACTORY_HOME/ on each Artifactory startup
  copyOnEveryStartup:
      - source: /tmp/artifactory-logback/*
        target: etc/
```

This change fix the permission issue.

But now i see the error. (Not a smooth journey ;))

 Failed initializing Home. Version 13 no found.

Any idea on why it is happening ?

**Comment by palko444** [comment]
Hello,

I have same issue. I'm installing from scratch 
error:
`kubectl logs pod/artifactory-ds-dev-artifactory-ha-primary-0 artifactory-ha
Copying bootstrap configs
cp: can't create '/artifactory_bootstrap/artifactory.config.import.xml': Permission denied`

command used for installation:
`helm upgrade --install --namespace $NAMESPACE $ARTIFACTORY_RELEASE_NAME jfrog/artifactory-ha  -f helm_out/values-artifactory.yaml  --version=2.4.6 --v 5 --wait --timeout 10m`

**Comment by jmcfallargo** [comment]
Like others - same/similar issue here: 
`Exec lifecycle hook ([/bin/bash -c echo; ]) for Container "artifactory-ha" in Pod "artifactory-ha-artifactory-ha-primary-0_artifactory-ha(38c1b130-5c4e-48a0-aedf-7fd273bf8ff4)" failed - error: command '/bin/bash -c echo; ' exited with 126: , message: "cannot exec in a stopped state: unknown\r\n"`

I am also having issues with customVolumes  and customVolumeMounts - if I specify a customVolume, I get an error that the configMap could not be found - verified that its not present.

This is all code that I took from a previous working installation, but now with upgraded helm (v3) and the latest chart.

I am currently blocked with deployment on this.

**Comment by chukka** [comment]
@palko444 @jmcfallargo  This is a known [issue ](https://github.com/jfrog/charts/issues/862#issuecomment-623567508), will be fixed as part of 7.5.x - ETA next week

**Comment by palko444** [comment]
@chukka i know its known issue, i did comment in the issue. I'm wondering how this can be open for ~23 days. Can you tell me what is last helm chart version not affected?
Thank you in advance

**Comment by chukka** [comment]
@palko444  this needed a fix in underlying docker image - hence it was planned for 7.5.0 release . If you are not using configMaps for bootstrapping  files , any prior version of 7.5.0 will work as expected,

**Comment by palko444** [comment]
Hello @chukka, I'm using configMaps, so what version can I use please?

**Comment by ripa1993** [comment]
I was able to get it running with latest chart version:

```yaml
customVolumes: |
  - name: import-config
    configMap:
      name: <config-map-name>
customVolumeMounts: |
  - name: import-config
     mountPath: "/artifactory_bootstrap/artifactory.config.import.xml"
     subPath: artifactory.config.import.xml
  - name: import-config
     mountPath: "/artifactory_bootstrap/security.import.xml"
     subPath: security.import.xml
```
instead of using
```yaml
configMapName: <config-map-name>
```

**Comment by kuritka** [comment]
Hi @chukka , for me the last version witout error is [2.0.19 FEB 26.](https://hub.helm.sh/charts/jfrog/artifactory-ha/2.0.19) From that time you made more than 40 releases. Are you sure the bug can't be anywhere else? i.e. somethng is wrong in configurations?

**Comment by jmcfallargo** [comment]
@kuritka I have tried too many versions now and I can't get Artifactory installed on any version - been on this for a week now -going to try 2.0.18

Does anybody know if its possible to use a new chart to install an old version of Artifactory?  We are still on version 6.x and I need some of the new features in the newer charts.

**Comment by chukka** [comment]
@ripa1993  Thanks for the workaround, @kuritka  you can try the workaround  but we  would like you to try the  permissions fix  in docker image (using configMapName ) in upcoming 7.5.0 Artifactory release . ETA 1 week :)

@jmcfallargo  new charts are only compatible with 7.x artifactory

**Comment by kuritka** [comment]
Hi @chukka  , I will try when new release comes but also need to deliver, so trying workarounds. 
I`m afraid workaround here is desinformative  as `/artifactory_extra_conf/{clusterlicense, info, binarystore.xml}` are not mentioned as well as `artifactory_extra_conf/logback.xml` . 

Can you provide correct configmap workaround ?

**Comment by kuritka** [comment]
@jmcfallargo @chukka @palko444 Hi, I resolved issue by updatig image versions in my local `values.yaml` and updating images in our onprem repo.

```Makefile
define copy-artifactory-to-harbor-v7.4.3
	docker pull docker.bintray.io/jfrog/nginx-artifactory-pro:7.4.3
	docker pull docker.bintray.io/jfrog/artifactory-pro:7.4.3
	docker pull postgres:12.3
	docker pull alpine:3.10
	docker pull busybox:1.30

	docker tag docker.bintray.io/jfrog/nginx-artifactory-pro:7.4.3 harbor.bison.prod.caas.absa.co.za/jfrog/nginx-artifactory-pro:7.4.3
	docker tag docker.bintray.io/jfrog/artifactory-pro:7.4.3 harbor.bison.prod.caas.absa.co.za/jfrog/artifactory-pro:7.4.3
	docker tag postgres:12.3 <our repo>/jfrog/postgres:12.3
	docker tag alpine:3.10 <our repo>/jfrog/alpine:3.10
	docker tag busybox:1.30 <our repo>/jfrog/busybox:1.30

	docker push harbor.bison.prod.caas.absa.co.za/jfrog/nginx-artifactory-pro:7.4.3
	docker push harbor.bison.prod.caas.absa.co.za/jfrog/artifactory-pro:7.4.3
	docker push <our repo>/jfrog/postgres:12.3
	docker push <our repo>/jfrog/alpine:3.10
	docker push <our repo>/jfrog/busybox:1.30
endef
```

**Comment by chukka** [comment]
@palko444 @angegar @kuritka  @jmcfallargo Finally, charts with fix  `artifactory-9.5.0, artifactory-ha-2.5.0 ` are released , please try and share the feedback

**Comment by palko444** [comment]
Hello,
bootstrap is ok now, however tomcat is not starting with this error

02-Jun-2020 12:02:00.881 INFO [main] org.apache.coyote.AbstractProtocol.init Initializing ProtocolHandler ["http-nio-8081"]
02-Jun-2020 12:02:00.903 INFO [main] org.apache.tomcat.util.net.NioSelectorPool.getSharedSelector Using a shared selector for servlet write/read
02-Jun-2020 12:02:00.917 INFO [main] org.apache.coyote.AbstractProtocol.init Initializing ProtocolHandler ["http-nio-127.0.0.1-8091"]
02-Jun-2020 12:02:00.918 INFO [main] org.apache.tomcat.util.net.NioSelectorPool.getSharedSelector Using a shared selector for servlet write/read
02-Jun-2020 12:02:00.920 INFO [main] org.apache.coyote.AbstractProtocol.init Initializing ProtocolHandler ["http-nio-127.0.0.1-8040"]
02-Jun-2020 12:02:00.921 INFO [main] org.apache.tomcat.util.net.NioSelectorPool.getSharedSelector Using a shared selector for servlet write/read
02-Jun-2020 12:02:00.930 INFO [main] org.apache.catalina.core.StandardService.startInternal Starting service [Catalina]
02-Jun-2020 12:02:00.930 INFO [main] org.apache.catalina.core.StandardEngine.startInternal Starting Servlet Engine: Apache Tomcat/8.5.54
02-Jun-2020 12:02:00.961 INFO [localhost-startStop-1] org.apache.catalina.startup.HostConfig.deployDescriptor Deploying deployment descriptor [/opt/jfrog/artifactory/app/artifactory/tomcat/conf/Catalina/localhos
t/access.xml]
02-Jun-2020 12:02:00.961 INFO [localhost-startStop-2] org.apache.catalina.startup.HostConfig.deployDescriptor Deploying deployment descriptor [/opt/jfrog/artifactory/app/artifactory/tomcat/conf/Catalina/localhos
t/artifactory.xml]
02-Jun-2020 12:02:01.014 WARNING [localhost-startStop-1] org.apache.catalina.startup.HostConfig.deployDescriptor A docBase [/opt/jfrog/artifactory/app/artifactory/tomcat/webapps/access.war] inside the host appBa
se has been specified, and will be ignored
02-Jun-2020 12:02:01.014 WARNING [localhost-startStop-2] org.apache.catalina.startup.HostConfig.deployDescriptor A docBase [/opt/jfrog/artifactory/app/artifactory/tomcat/webapps/artifactory.war] inside the host 
appBase has been specified, and will be ignored
02-Jun-2020 12:02:08.438 INFO [localhost-startStop-2] org.apache.catalina.startup.HostConfig.deployDescriptor Deployment of deployment descriptor [/opt/jfrog/artifactory/app/artifactory/tomcat/conf/Catalina/loca
lhost/artifactory.xml] has finished in [7,475] ms
02-Jun-2020 12:02:18.176 INFO [localhost-startStop-1] org.apache.catalina.startup.HostConfig.deployDescriptor Deployment of deployment descriptor [/opt/jfrog/artifactory/app/artifactory/tomcat/conf/Catalina/loca
lhost/access.xml] has finished in [17,214] ms
02-Jun-2020 12:02:18.177 INFO [localhost-startStop-2] org.apache.catalina.startup.HostConfig.deployDirectory Deploying web application directory [/opt/jfrog/artifactory/app/artifactory/tomcat/webapps/ROOT]
02-Jun-2020 12:02:18.201 INFO [localhost-startStop-2] org.apache.catalina.startup.HostConfig.deployDirectory Deployment of web application directory [/opt/jfrog/artifactory/app/artifactory/tomcat/webapps/ROOT] h
as finished in [24] ms
02-Jun-2020 12:02:18.217 INFO [main] org.apache.coyote.AbstractProtocol.start Starting ProtocolHandler ["http-nio-8081"]
02-Jun-2020 12:02:18.290 INFO [main] org.apache.coyote.AbstractProtocol.start Starting ProtocolHandler ["http-nio-127.0.0.1-8091"]
02-Jun-2020 12:02:18.347 INFO [main] org.apache.coyote.AbstractProtocol.start Starting ProtocolHandler ["http-nio-127.0.0.1-8040"]
02-Jun-2020 12:06:03.823 INFO [main] org.apache.catalina.core.StandardServer.await A valid shutdown command was received via the shutdown port. Stopping the Server instance.

**Comment by chukka** [comment]
@palko444 Can you please raise this as a new issue (also share reproducible steps )

**Comment by kuritka** [comment]
@palko444 , sorry I jump in. Can you share output of this command and version you installed? 

```
kubectl get po -n <artifactory namespace>  -o yaml | grep image: | cut -d ":" -f2,3 | sort | uniq
```

**Comment by palko444** [comment]
Hello @kuritka,

alpine:3.10
 bitnami/minideb:stretch
 bitnami/postgres-exporter:0.8.0-debian-10-r80
 bitnami/postgresql:11.7.0-debian-10-r9
 docker.bintray.io/jfrog/artifactory-pro:7.5.5
 docker.io/bitnami/minideb:stretch
 docker.io/bitnami/postgres-exporter:0.8.0-debian-10-r80
 docker.io/bitnami/postgresql:11.7.0-debian-10-r9
 fluent/fluent-bit:1.4
 quay.io/travelaudience/docker-nexus:3.21.2-03

**Comment by kuritka** [comment]
@chukka Hi, seems, 7.5.0 is working well for me . Anyway, as I mentioned. My struggles was with incorrect container versions. Not artifactory version itself. 
Another issue I had with 7.4.3 - I could't join with xray telling me `Cluster join: Failed resolving join key: failed resolving 'shared.security.joinKey' key; file does not exist: /opt/jfrog/artifactor`. Upgrade to  7.5.5 helps (I think problem in 7.4.3 was with race condition because I have disabled `waitForPrimaryStartup`). Till now I can tell 7.5.5 is working as expected

**Comment by kuritka** [comment]
Hi @palko444 look at my containres in dedicated artifactory namespace (nothing else than artifacotry is there)

```
.../alpine:3.11
 .../artifactory-pro:7.5.5
 .../nginx-artifactory-pro:7.5.5
 .../postgres:9.6.18-debian-10-r7
```
Can you download and unzip proper helmchart from [here](https://charts.jfrog.io/) , look into `values.yaml` and fix container versions in your local `values.yaml` ? I think your artifactory will work after you fix container versions.

**Comment by palko444** [comment]
Hello @kuritka,
i was missing some "tomcat" lines in my values file, im not sure where this was added. Anyway bigger issue is remove of running java "keytool" in "entry-point.sh" script. As i have custom cert for S3 artifactory failed to start even if I had certs mounted in "artifactory_extra_certs" folder. I checked entry-point.sh and these certs are no more added to java key store.
All is ok after I added **preStartCommand**: "/opt/jfrog/artifactory/app/third-party/java/bin/keytool -importcert -keystore ...."
thx for help
